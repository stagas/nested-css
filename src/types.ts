// TODO: docs are not rendering correctly for this type
// https://github.com/TypeStrong/typedoc/issues/1519
/**
 * Nested CSS rules.
 */
export type NestedCSSDeclaration = Partial<CSSStyleDeclaration> & {
  [name: string]: string | NestedCSSDeclaration
}
