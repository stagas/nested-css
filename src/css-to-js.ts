import { createTokenizer } from 'tokenizer-next'
import type { NestedCSSDeclaration } from './types'

const extraWhitespace = /\s+/g

const tokenizer = createTokenizer(
  /(?<open>\s*\{\s*)/,
  /(?<close>\s*\}\s*)/,
  /(?<comment>\/\*[^]*?\*\/)/,
  /\s*(?<rule>[^{;}/'"]+)\s+?(?={)/,
  /\s*(?<string>'.*?'|".*?")(?=;)/,
  /\s*(?<prop>[^:;/]+)(?=:)/,
  /\s*(?<value>[^:;]+)(?=;)/,
)

/**
 * Convert a CSS string to a {@link NestedCSSDeclaration}.
 *
 * @param input
 */
export function cssToJs(input: string) {
  input = input.replace(extraWhitespace, ' ')

  const nextToken = tokenizer(input)
  const filter = (ignored: string[]) => (() => {
    let token
    while (token = nextToken())
      if (!ignored.includes(token.group)) break
    return token
  })
  const next = filter(['comment'])

  const parse = (style: NestedCSSDeclaration = {}) => {
    let token

    while ((token = next())) {
      const { value, group } = token

      switch (group) {
        case 'rule':
          next() // open {
          // @ts-ignore can't handle the ??= for some reason
          parse(style[value.trim()] ??= {})
          break
        case 'prop':
          style[value.trim()] = (next()?.value as string)?.trim()
          break
        case 'close':
          return style
      }
    }

    return style
  }

  return parse()
}
