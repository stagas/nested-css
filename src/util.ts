/**
 * Joins parts with values
 *
 * @param parts
 * @param values
 */
export function joinPartsWithValues(parts: TemplateStringsArray, values: unknown[]) {
  let str = ''
  for (let i = 0; i < parts.length; i++) {
    str += parts[i]
    // if (i < values.length) str += values[i]
    str += values[i]
  }
  return str
}

/**
 * Convert string to kebab-case.
 *
 * @param s PascalCase or camelCase string
 */
export function kebabCase(s: string) {
  return s.replace(/[A-Z]/g, (m: string, i: number) => (i ? '-' : '') + m).toLowerCase()
}
