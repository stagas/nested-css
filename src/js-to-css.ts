// original: https://gist.github.com/mindplay-dk/47bc597dcbc28bea009b02c5a0c9010e

import type { NestedCSSDeclaration } from './types'
import { kebabCase } from './util'

const mediaRule = (rule: string, media: string) => (media ? `${media}{${rule}}` : rule)

const createRule = (target: string, prop: string, value: string) => `${target}{${kebabCase(prop)}:${value}}`

// const isClassOrRoot = /^\s*[.:]/g

const remap = (target: string, map?: Map<string, string>) =>
  // isClassOrRoot.test(target)
  //   ? target
  //   :
  [...target.matchAll(/[\w-]+|./g)]
    .flat()
    .map(x => map?.get(x) ?? x)
    .join('')

/**
 * Compile a JS nested rules {@link NestedCSSDeclaration} to a CSS string.
 *
 * Examples:
 * ```ts
 * jsToCss({ '.foo': { color: 'blue' } })
 * // => .foo{color:blue}
 *
 * // custom root
 * jsToCss({ color: 'red' }, '.my-button')
 * // => .my-button{color:red}
 *
 * // with substitution
 * jsToCss({ '.foo': { color: 'blue' } }, null, new Map([['foo', 'bar']]))
 * // => .bar{color:blue}
 * ```
 *
 * @param rules Rules object
 * @param rootSelector Top level rules will use this selector
 * @param aliasMap Alias identifiers (i.e for `.foo` to become `.bar`
 *  you will need a `foo=bar` entry)
 * @returns The compiled CSS string
 */
export function jsToCss(
  rules: NestedCSSDeclaration,
  rootSelector?: string | null | undefined,
  aliasMap?: Map<string, string>,
) {
  rootSelector ??= ''

  let css = ''

  const parse = (rules: NestedCSSDeclaration, child = '', media = '') => {
    for (const key in rules) {
      const value = rules[key]

      if (typeof value === 'object') {
        const isMedia = /^@/.test(key) ? key : null

        let target = child

        if (!isMedia) {
          target = key
            .split(',')
            .map(k => k.trim())
            .map(k =>
              child
                .split(',')
                .map(x => (x + (/^&/.test(k) ? k.slice(1) : ' ' + k)).trim())
                .join(',')
            )
            .map(k => k.trim())
            .join(',')
        }

        parse(value, target, isMedia || media)
      } else {
        const rule = mediaRule(
          createRule(child ? remap(child, aliasMap) : rootSelector!, key, value),
          media,
        )
        css += '\n' + rule.trim()
      }
    }
  }

  parse({
    [rootSelector]: rules,
  } as NestedCSSDeclaration)

  return css.trim()
}
