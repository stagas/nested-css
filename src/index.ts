export type { NestedCSSDeclaration } from './types'

import { cssToJs } from './css-to-js'
import { jsToCss } from './js-to-css'
import { joinPartsWithValues } from './util'

/**
 * Compile to CSS passing parameters to {@link jsToCss}.
 */
export type NestedCSSCompiler = ((
  rootSelector?: string,
  aliasMap?: Map<string, string>
) => string) & { valueOf: NestedCSSCompiler }

/**
 * Factory a {@link NestedCSSCompiler} for the given string.
 *
 * @param parts
 * @param values
 */
export function css(parts: TemplateStringsArray, ...values: unknown[]): NestedCSSCompiler {
  const nestedCssString = joinPartsWithValues(parts, values)
  const nestedCssObject = cssToJs(nestedCssString)

  function compileCss(rootSelector = ':host', aliasMap?: Map<string, string>) {
    return jsToCss(nestedCssObject, rootSelector, aliasMap)
  }

  compileCss.valueOf = compileCss

  return compileCss
}

export { cssToJs, jsToCss }
