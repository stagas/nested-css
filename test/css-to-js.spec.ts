import { cssToJs } from '../src/css-to-js'

describe('css to js', () => {
  it('should parse css to js object', () => {
    const result = cssToJs(`
      color: magenta;

      :root {
        color: blue;
        background: green;
        /* comment */
      }
      /* comment */
      .foo,
      .bar {
        color : red;

        &:hover {
          color: yellow;
        }
      }

      deep .selector:hover foo::after {
        margin: 0 auto;

        .deep {
          .nested {
            &:focus {
              visibility: hidden;
            }
          }
        }

        content: 'continues';
      }
    `)

    expect(result).toEqual({
      color: 'magenta',

      ':root': {
        color: 'blue',
        background: 'green',
      },

      '.foo, .bar': {
        color: 'red',

        '&:hover': {
          color: 'yellow',
        },
      },

      'deep .selector:hover foo::after': {
        margin: '0 auto',

        '.deep': {
          '.nested': {
            '&:focus': {
              visibility: 'hidden',
            },
          },
        },

        content: '\'continues\'',
      },
    })
  })

  it('should not break on broken css prop', () => {
    const result = cssToJs(`
      color:
    `)
    expect(result).toEqual({ color: undefined })
  })

  it('cascade', () => {
    const result = cssToJs(`
      .name {
        > span {
          color: var(--blue);
        }
        position: absolute;
      }
      .name {
        > span {
          background: red;
        }
        display: block;
      }
    `)
    expect(result).toEqual({
      '.name': {
        '> span': {
          color: 'var(--blue)',
          background: 'red',
        },
        position: 'absolute',
        display: 'block',
      },
    })
  })

  it('should ignore string content', () => {
    const result = cssToJs(`
      content: ':;\\'"';
      color: blue;
      .other {
        content: ":;\\"'";
        color: blue;
      }
    `)
    expect(result).toEqual({
      content: '\':;\\\'"\'',
      color: 'blue',
      '.other': {
        content: '":;\\"\'"',
        color: 'blue',
      },
    })
  })

  it('should parse root &', () => {
    const result = cssToJs(`
      color: magenta;
      & {
        color: yellow;
        .another {
          color: blue;
        }
      }
    `)
    expect(result).toEqual({
      color: 'magenta',
      '&': { color: 'yellow', '.another': { color: 'blue' } },
    })
  })

  it('comments in strings', () => {
    const result = cssToJs(`
      .short-text {
        font-style: italic;
        &:before {
          content: '/* ';
        }
        &:after {
          content: '*/ ';
        }
      }
    `)
    expect(result).toMatchSnapshot()
  })

  it('comments case 2', () => {
    const result = cssToJs(`
      .info {
        > .type {
          display: inline-block;
          padding: 0;
          vertical-align: top;
          /*&:before {
            content: ': ';
            color: var(--red);
          }*/
        }
      }
    `)
    expect(result).toMatchSnapshot()
  })

  it('case brackets', () => {
    const result = cssToJs(`
      &:after {
        content: ' {';
        color: var(--white);
      }
      &:before {
        content: ' }';
        color: var(--white);
      }
    `)
    expect(result).toMatchSnapshot()
  })
})
