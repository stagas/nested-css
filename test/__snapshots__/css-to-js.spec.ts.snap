// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css to js comments in strings 1`] = `
Object {
  ".short-text": Object {
    "&:after": Object {
      "content": "'*/ '",
    },
    "&:before": Object {
      "content": "'/* '",
    },
    "font-style": "italic",
  },
}
`;
